### PR TITLE
Update to egui `0.29.0` & Enhancements

### PR DESCRIPTION
**Pull request by bircni**

### CODE REVIEW

**Comment by bircni** [comment]
Just noticed
In egui_flex are many unused variables and dead code - I will remove them in a seperate commit - please tell me if you want that, else I will remove this commit then

**Comment by lucasmerlin** [comment]
Awesome, thank you 🙏

Yeah I need to do a final clean up / documentation pass before I publish egui_flex, I plan to publish it next week. If you want to clean it up beforehand, that would be appreciated!

**Comment by bircni** [comment]
i can try :-)

What about linting?
Should I enable some lints too?

File: crates/egui_flex/src/lib.rs [hunk 1]
```diff
@@ -824,37 +733,23 @@ impl FlexContainerUi {
...
+            ..
         } = self;
 
         // We will assume that the margin is symmetrical
         let margin_top_left = ui.min_rect().min - frame_rect.min;
 
         let child_rect = content_rect.intersect(ui.max_rect());
 
-        let mut child = ui.child_ui(child_rect, *ui.layout(), None);
+        let mut child = ui.new_child(UiBuilder::new().max_rect(child_rect).layout(*ui.layout()));
```
Comment (lucasmerlin) [inline crates/egui_flex/src/lib.rs:747]: remove .layout

File: crates/egui_router/src/transition.rs [hunk 2]
```diff
@@ -16,7 +16,12 @@ pub trait ComposableTransitionTrait {
 
 impl<T: ComposableTransitionTrait> TransitionTrait for T {
     fn create_child_ui(&self, ui: &mut Ui, t: f32, with_id: Id) -> Ui {
-        let mut child = ui.child_ui_with_id_source(ui.max_rect(), *ui.layout(), with_id, None);
+        let mut child = ui.new_child(
+            UiBuilder::new()
+                .max_rect(ui.max_rect())
+                .layout(*ui.layout())
```
Comment (lucasmerlin) [inline crates/egui_router/src/transition.rs:22]: layout

File: crates/egui_router/src/transition.rs [hunk 3]
```diff
@@ -96,7 +101,12 @@ impl TransitionTrait for SlideTransition {
         let offset = available_size * (1.0 - t) * self.amount;
         let child_rect = ui.max_rect().translate(offset);
 
-        ui.child_ui_with_id_source(child_rect, *ui.layout(), with_id, None)
+        ui.new_child(
+            UiBuilder::new()
+                .max_rect(child_rect)
+                .layout(*ui.layout())
```
Comment (lucasmerlin) [inline crates/egui_router/src/transition.rs:107]: layout

File: crates/egui_router/src/transition.rs [hunk 4]
```diff
@@ -200,7 +210,7 @@ impl ActiveTransition {
         &mut self,
         ui: &mut Ui,
         state: &mut State,
-        (in_id, content_in): (usize, impl FnOnce(&mut Ui, &mut State)),
+        (in_id, contentin_): (usize, impl FnOnce(&mut Ui, &mut State)),
```
Comment (lucasmerlin) [inline crates/egui_router/src/transition.rs:213]: Find and replace mistake? 🤔

File: crates/egui_router/src/transition.rs [hunk 5]
```diff
@@ -261,11 +271,11 @@ impl ActiveTransition {
...
         with_temp_auto_id(ui, with_id, |ui| {
-            let mut ui = ui.child_ui_with_id_source(
-                ui.max_rect(),
-                *ui.layout(),
-                Id::new("router_child").with(with_id),
-                None,
+            let mut ui = ui.new_child(
+                UiBuilder::new()
+                    .max_rect(ui.max_rect())
+                    .layout(*ui.layout())
```
Comment (lucasmerlin) [inline crates/egui_router/src/transition.rs:277]: layout

File: crates/egui_virtual_list/src/lib.rs [hunk 6]
```diff
@@ -169,7 +170,11 @@ impl VirtualList {
         // to maintain the scroll position
         let scroll_items_top_step_2 =
             if let Some(scroll_top_items) = self.items_inserted_at_start.take() {
-                let mut measure_ui = ui.child_ui(ui.max_rect(), *ui.layout(), None);
+                let mut measure_ui = ui.new_child(
+                    UiBuilder::new()
+                        .max_rect(ui.max_rect())
+                        .layout(*ui.layout()),
```
Comment (lucasmerlin) [inline crates/egui_virtual_list/src/lib.rs:176]: layout

File: crates/hello_egui_utils/src/center.rs [hunk 7]
```diff
@@ -48,7 +48,7 @@ impl Center {
             rect
         };
 
-        let mut ui = ui.child_ui(content_rect, *ui.layout(), None);
+        let mut ui = ui.new_child(UiBuilder::new().max_rect(content_rect).layout(*ui.layout()));
```
Comment (lucasmerlin) [inline crates/hello_egui_utils/src/center.rs:51]: layout

File: fancy-example/src/chat.rs [hunk 8]
```diff
@@ -211,7 +213,11 @@ impl ChatExample {
                                 // chat bubble layout where the own bubbles are right-aligned and
                                 // the text within is left-aligned.
                                 let (_pos, galley, _response) = label.layout_in_ui(
-                                    &mut ui.child_ui(ui.max_rect(), *ui.layout(), None),
+                                    &mut ui.new_child(
+                                        UiBuilder::new()
+                                            .max_rect(ui.max_rect())
+                                            .layout(*ui.layout()),
```
Comment (lucasmerlin) [inline fancy-example/src/chat.rs:219]: layout

File: fancy-example/src/chat.rs [hunk 9]
```diff
@@ -273,25 +279,25 @@ impl ChatExample {
...
+                                let top = response.rect.left_top() + Vec2::splat(margin);
                                 let arrow_rect =
-                                    Rect::from_two_pos(top, top + Vec2::new(rounding, rounding));
+                                    Rect::from_two_pos(top, top + Vec2::new(-rounding, rounding));
 
                                 vec![
                                     arrow_rect.left_top(),
                                     arrow_rect.right_top(),
-                                    arrow_rect.left_bottom(),
+                                    arrow_rect.right_bottom(),
```
Comment (lucasmerlin) [inline fancy-example/src/chat.rs:300]: What does this change?

**Review by lucasmerlin — APPROVED** [review]
Awesome, thank you so much!

**Comment by Ruttie2006** [comment]
Since this PR is arbitrarily named 'enhancements', would adding #36 to this pr be possible?

**Comment by lucasmerlin** [comment]
> Since this PR is arbitrarily named 'enhancements', would adding https://github.com/lucasmerlin/hello_egui/issues/36 to this pr be possible?

I'd prefere a separate one since this one is already reeeealy big

File: Cargo.toml [hunk 10]
```diff
@@ -65,9 +82,33 @@ egui_infinite_scroll = { path = "./crates/egui_infinite_scroll", version = "0.4.
...
+egui_extras = { version = "0.29.0", default-features = false }
+
+[workspace.lints.rust]
+unsafe_code = "deny"
+
+[workspace.lints.clippy]
+pedantic = { level = "deny", priority = 1 }
+enum_glob_use = { level = "deny", priority = 2 }
+perf = { level = "warn", priority = 3 }
+style = { level = "warn", priority = 4 }
```
[the remark this answers is not in the document]
Comment (bircni) [reply Cargo.toml:96]: https://rust-lang.github.io/rust-clippy/master/index.html#/lint_groups_priority

File: fancy-example/src/chat.rs (hunk 9 above)
Comment (bircni) [reply fancy-example/src/chat.rs:300]: Just swapped the if else to make it easier to raead (clippy told me to!)

**Comment by bircni** [comment]
> I'd prefere a separate one since this one is already reeeeally big

Ooops xD

**Comment by bircni** [comment]
Now everything should be ready to merge :-)

https://github.com/bircni/hello_egui/actions/runs/11087433518

**Comment by bircni** [comment]
Maybe we could also set all versions on the same version number as egui is. So it has the same number and all crate versions here can be managed by the main cargo.toml file with the workspace

File: fancy-example/src/chat.rs (hunk 9 above)
Comment (lucasmerlin) [reply fancy-example/src/chat.rs:296]: Aaah I see, the way the diff is shown is a bit confusing

**Comment by lucasmerlin** [comment]
> Maybe we could also set all versions on the same version number as egui is

I actually thought about that before but decided against it because it would mean I can't release a breaking change to a crate until the next egui release. For now I'm fine with separate versions ([cargo release](https://github.com/crate-ci/cargo-release) makes this really easy to manage)

**Review by lucasmerlin — APPROVED** [review]

**Comment by bircni** [comment]
Do you need something from me or is it done then?😅

**Comment by lucasmerlin** [comment]
It's done, I was just waiting on the pipeline 😄

**Pull request merged**